File: flutter/lib/DBBenchmark.dart
import 'dart:math';

import 'package:path/path.dart';
import 'package:sqflite/sqflite.dart';

import 'helper/User.dart';

class DBBenchmark {
  static const int _times = 1000;
  static const int _tableSize = 10000;
  int _timer = 0;

  Database _db;

  DBBenchmark() {
    _openDb();
  }

  void _openDb() async {
    _db = await openDatabase(join(await getDatabasesPath(), 'db_flutter.db'),
        onCreate: (db, version) {
      return db.execute(DB.SQL_CREATE_TABLE);
    }, version: 1);
  }

  Future<double> insert() async {
    int result = 0;

    await _clearTable();
    User user;
    for (int i = 0; i < _times; i++) {
      user = _dummyUser(i);
      _timer = DateTime.now().microsecondsSinceEpoch;
      await _insert(user);
      result += DateTime.now().microsecondsSinceEpoch - _timer;
    }

    return result / _times;
  }

  Future<double> get() async {
    int result = 0;

    await _prepareData(_tableSize);
    int dummy = 0;
    int id;
    for (int i = 0; i < _times; i++) {
      id = Random().nextInt(_tableSize);
      _timer = DateTime.now().microsecondsSinceEpoch;
      dummy += (await _get(id)).id;
      result += DateTime.now().microsecondsSinceEpoch - _timer;
    }

    print(dummy);
    return result / _times;
  }

  Future<double> getAll() async {
    int result = 0;

    await _prepareData(_tableSize);
    int dummy = 0;
    for (int i = 0; i < _times / 20; i++) {
      _timer = DateTime.now().microsecondsSinceEpoch;
      dummy += (await _getAll()).length;
      result += DateTime.now().microsecondsSinceEpoch - _timer;
    }

    print(dummy);
    return result / (_times / 20);
  }

  Future<double> update() async {
    int result = 0;

    await _prepareData(_tableSize);
    int id;
    for (int i = 0; i < _times; i++) {
      id = Random().nextInt(_tableSize);
      _timer = DateTime.now().microsecondsSinceEpoch;
      await _update(id, i);
      result += DateTime.now().microsecondsSinceEpoch - _timer;
    }

    return result / _times;
  }

  Future<double> delete() async {
    int result = 0;

    await _prepareData(_tableSize + _times);
    int id;
    for (int i = 0; i < _times; i++) {
      id = Random().nextInt(_tableSize);
      _timer = DateTime.now().microsecondsSinceEpoch;
      await _delete(id);
      result += DateTime.now().microsecondsSinceEpoch - _timer;
    }

    return result / _times;
  }

  Future<void> _insert(User user) async {
    await _db.insert(DB.TABLE_NAME, _toMap(user));
  }

  Future<User> _get(int id) async {
    var query = await _db.query(DB.TABLE_NAME, where: DB.COLUMN_ID + ' = ?', whereArgs: [id]);
    return query.isNotEmpty ? _fromMap(query.first) : null;
  }

  Future<List<User>> _getAll() async {
    return (await _db.query(DB.TABLE_NAME)).map((q) => _fromMap(q)).toList();
  }

  Future<void> _update(int id, int i) async {
    return await _db.update(DB.TABLE_NAME, {DB.COLUMN_NAME: id + i}, where: DB.COLUMN_ID + ' = ?', whereArgs: [id]);
  }

  Future<void> _delete(int id) async {
    return await _db.delete(DB.TABLE_NAME, where: DB.COLUMN_ID + ' = ?', whereArgs: [id]);
  }

  Future<void> _clearTable() async {
    await _db.execute(DB.SQL_CLEAR_TABLE);
  }

  Future<void> _prepareData(int size) async {
    await _clearTable();

    var batch = _db.batch();
    for (int i = 0; i < size; i++) {
      batch.insert(DB.TABLE_NAME, _toMap(_dummyUser(i)));
    }
    await batch.commit(noResult: true);
  }

  User _dummyUser(int i) {
    return User(i, "login $i", "email $i", "name $i", 30);
  }

  User _fromMap(Map<String, dynamic> query) {
    return User(query[DB.COLUMN_ID], query[DB.COLUMN_LOGIN], query[DB.COLUMN_EMAIL], query[DB.COLUMN_NAME], query[DB.COLUMN_AGE]);
  }

  Map<String, dynamic> _toMap(User user) {
    return <String, dynamic>{
      DB.COLUMN_ID: user.id,
      DB.COLUMN_LOGIN: user.login,
      DB.COLUMN_EMAIL: user.email,
      DB.COLUMN_NAME: user.name,
      DB.COLUMN_AGE: user.age
    };
  }
}

class DB {
  static const String TABLE_NAME = "USERS";
  static const String COLUMN_ID = "ID";
  static const String COLUMN_LOGIN = "LOGIN";
  static const String COLUMN_EMAIL = "EMAIL";
  static const String COLUMN_NAME = "NAME";
  static const String COLUMN_AGE = "AGE";

  static const String SQL_CREATE_TABLE =
      "CREATE TABLE $TABLE_NAME (" +
          "$COLUMN_ID INTEGER PRIMARY KEY, " +
          "$COLUMN_LOGIN TEXT UNIQUE, " +
          "$COLUMN_EMAIL TEXT UNIQUE, " +
          "$COLUMN_NAME TEXT NOT NULL, " +
          "$COLUMN_AGE INTEGER NOT NULL CHECK ($COLUMN_AGE > 0 AND $COLUMN_AGE < 150))";

  static const String SQL_CLEAR_TABLE = "DELETE FROM $TABLE_NAME";
}

File: kotlin/app/src/commonMain/kotlin/kotlin_test/DBBenchmark.kt
package kotlin_test

import com.squareup.sqldelight.db.SqlDriver
import kotlin_test.data.UserQueries
import kotlin.random.Random

class DBBenchmark(driver: SqlDriver) {

    private val times = 1000
    private val tableSize = 10000
    private var timer = 0L

    private val userQueries: UserQueries = TestDB(driver).userQueries

    fun insert(): Double {
        var result = 0L

        clearTable()
        var user: User
        repeat(times) {
            user = dummyUser(it)
            timer = Util.getNanoTime()
            insert(user)
            result += Util.getNanoTime() - timer
        }

        return result / times.toDouble()
    }

    fun get(): Double {
        var result = 0L

        prepareData(tableSize)
        var dummy = 0L
        var id: Long
        repeat(times) {
            id = Random.nextLong(tableSize.toLong())
            timer = Util.getNanoTime()
            dummy += get(id).id
            result += Util.getNanoTime() - timer
        }

        println(dummy)
        return result / times.toDouble()
    }

    fun getAll(): Double {
        var result = 0L

        prepareData(tableSize)
        var dummy = 0
        repeat(times / 20) {
            timer = Util.getNanoTime()
            dummy += findAll().size
            result += Util.getNanoTime() - timer
        }

        println(dummy)
        return result / (times / 20.0)
    }

    fun update(): Double {
        var result = 0L

        prepareData(tableSize)
        var id: Long
        repeat(times) {
            id = Random.nextLong(tableSize.toLong())
            timer = Util.getNanoTime()
            update(id, it)
            result += Util.getNanoTime() - timer
        }

        return result / times.toDouble()
    }

    fun delete(): Double {
        var result = 0L

        prepareData(tableSize + times)
        var id: Long
        repeat(times) {
            id = Random.nextLong(tableSize.toLong())
            timer = Util.getNanoTime()
            delete(id)
            result += Util.getNanoTime() - timer
        }

        return result / times.toDouble()
    }

    private fun insert(user: User) = userQueries.insert(user.id, user.login, user.email, user.name, user.age)

    private fun get(id: Long): User = userQueries.selectOne(id).executeAsOne().toUser()

    private fun findAll(): List<User> = userQueries.selectAll().executeAsList().map { it.toUser() }

    private fun update(id: Long, i: Int) = userQueries.updateName((id + i).toString(), id)

    private fun delete(id: Long) = userQueries.delete(id)

    private fun clearTable() = userQueries.clearTable()

    private fun prepareData(size: Int) {
        clearTable()
        repeat(size) {
            insert(dummyUser(it))
        }
    }

    private fun dummyUser(i: Int) = User(i.toLong(), "login $i", "email $i", "name &$i", 30)

}